src/ItemTreeTraversal/MysqlItemTreeTraversal.php

```diff
@@ -53,18 +58,56 @@ public function getParentRows(string $name): array
 
     public function getChildrenRows(string $name): array
     {
-        $sql = "SELECT DISTINCT item.* FROM (
-            SELECT DISTINCT child
-            FROM (SELECT * FROM $this->childrenTableName ORDER by parent) item_child_sorted,
-            (SELECT @pv := :name) init
-            WHERE find_in_set(parent, @pv) AND length(@pv := concat(@pv, ',', child))
-        ) s
-        LEFT JOIN $this->tableName AS item ON item.name = s.child";
+        $baseOuterQuery = (new Query($this->database))->select([new Expression('item.*')])->distinct();
 
         /** @psalm-var RawItem[] */
-        return $this
-            ->database
-            ->createCommand($sql, [':name' => $name])
-            ->queryAll();
+        return $this->getChildrenRowsCommand($name, baseOuterQuery: $baseOuterQuery)->queryAll();
+    }
+
+    public function getChildPermissionRows(string $name): array
+    {
+        $baseOuterQuery = (new Query($this->database))
+            ->select([new Expression('item.*')])
+            ->distinct()
+            ->where(['item.type' => Item::TYPE_PERMISSION]);
+
+        /** @psalm-var RawItem[] */
+        return $this->getChildrenRowsCommand($name, baseOuterQuery: $baseOuterQuery)->queryAll();
+    }
+
+    public function getChildRoleRows(string $name): array
+    {
+        $baseOuterQuery = (new Query($this->database))
+            ->select([new Expression('item.*')])
+            ->distinct()
+            ->where(['item.type' => Item::TYPE_ROLE]);
+
+        /** @psalm-var RawItem[] */
+        return $this->getChildrenRowsCommand($name, baseOuterQuery: $baseOuterQuery)->queryAll();
+    }
+
+    public function hasChild(string $parentName, string $childName): bool
+    {
+        $baseOuterQuery = (new Query($this->database))
+            ->select([new Expression('1 AS item_child_exists')])
+            ->andWhere(['item.name' => $childName]);
+        /** @psalm-var array<0, 1>|false $result */
+        $result = $this->getChildrenRowsCommand($parentName, baseOuterQuery: $baseOuterQuery)->queryScalar();
+
+        return $result !== false;
+    }
+
+    private function getChildrenRowsCommand(string $name, QueryInterface $baseOuterQuery): CommandInterface
+    {
+        $fromSql = "SELECT DISTINCT child
+        FROM (SELECT * FROM $this->childrenTableName ORDER by parent) item_child_sorted,
+        (SELECT @pv := :name) init
+        WHERE find_in_set(parent, @pv) AND length(@pv := concat(@pv, ',', child))";
+        $outerQuery = $baseOuterQuery
+            ->from(new Expression("($fromSql) s"))
```

Comment:
```suggestion
            ->from("($fromSql) s")
```

Reply:
The tests are not passing:

- https://github.com/yiisoft/rbac-db/pull/46/commits/1df2b5e289bfb34ba554baac2361d964796986ad
- https://github.com/yiisoft/rbac-db/actions/runs/6295365620/job/17088515626

Reply:
Tests passed for latest MySQL version https://github.com/yiisoft/rbac-db/actions/runs/6295365620/job/17095505753

Why it depends on MySQL version?

Could you try this `->from(['s' => "($fromSql)"])`?

Reply:
The latter worked, thanks.